lib/main.js

```diff
@@ -1,34 +1,36 @@
 'use strict'
 
 var dotenvExpand = function (config) {
-  var interpolate = function (env) {
-    var matches = env.match(/\$([a-zA-Z0-9_]+)|\${([a-zA-Z0-9_]+)}/g) || []
-
-    matches.forEach(function (match) {
-      var key = match.replace(/\$|{|}/g, '')
-
-      // process.env value 'wins' over .env file's value
-      var variable = process.env.hasOwnProperty(key) ? process.env[key] : (config.parsed[key] || '')
-
-      // Resolve recursive interpolations
-      variable = interpolate(variable)
-
-      env = env.replace(match, variable)
-    })
-
-    return env
+  var interpolate = function (envValue) {
+    var matches = envValue.match(/(.?\${?(?:[a-zA-Z0-9_]+)?}?)/g) || []
+
+    return matches.reduce(function (newEnv, match) {
+      var parts = /(.?)\${?([a-zA-Z0-9_]+)?}?/g.exec(match)
+      var $prefix = parts[1]
+
+      var value, replacePart
+
+      if ($prefix === '\\') {
+        replacePart = parts[0]
+        value = replacePart.replace('\\$', '$')
+      } else {
+        var key = parts[2]
+        replacePart = parts[0].substring($prefix.length)
+        // process.env value 'wins' over .env file's value
+        value = process.env.hasOwnProperty(key) ? process.env[key] : (config.parsed[key] || '')
+
+        // Resolve recursive interpolations
+        value = interpolate(value)
+      }
+
+      return newEnv.replace(replacePart, value)
+    }, envValue)
   }
 
   for (var configKey in config.parsed) {
     var value = process.env.hasOwnProperty(configKey) ? process.env[configKey] : config.parsed[configKey]
 
-    if (config.parsed[configKey].substring(0, 2) === '\\$') {
-      config.parsed[configKey] = value.substring(1)
-    } else if (config.parsed[configKey].indexOf('\\$') > 0) {
-      config.parsed[configKey] = value.replace(/\\\$/g, '$')
-    } else {
-      config.parsed[configKey] = interpolate(value)
-    }
+    config.parsed[configKey] = interpolate(value)
```

Comment:
It's nice how this simplifies the logic here to a single method method call.

Reply:
👍